<h1 style="font-family: Baskerville; color: red; font-size: 24px; font-weight: normal; padding: 10px; line-height:30px;">Topic 7: OOP in general and accessing Python&rsquo;s objects/variables.  <br />
    (Optional Appendix).
</h1>
<p style="color:white;">&copy; GB 2023 Feb 18</p>
<hr />
<blockquote>
    Be sure to install colorama [pip install colorama].  See <a href="https://pypi.org/project/colorama/" target="new">https://pypi.org/project/colorama/</a>
</blockquote>

<hr />
<h2 style="color:cornflowerblue;">Singly-Linked List demo</h2>
<p>This isn't quite finished but ... the singly-linked list is a dynamic data structure consisting of a set of <b>nodes</b>.  Each node has two parts: a data value and a pointer or reference to the next pointer in a linear sequence. E.g., <pre>[Head|->] [5|->node_at_y1] [7|->node_at_x2] [Tail|->NULL</pre>
</p>
<p>The only reason this is demo'd here is show that we can store references to our nodes as objects in a list and then pop them off the list and reconstruct them as Nodes to extract data and references.</p>
<p>The linkedList1.py file is in our weekly folder.</p>

In [3]:
%%bash
python linkedlist1.py





--------------------------------------------------

Welcome to the Singly Linked List Q&A
	We start off with no nodes ... if none, create one and start at the Head.
	Notice new nodes are pushed at the head; they're popped from the head, too.
	A node is shown here with two parts [value|•>pointer 	to next node].



	[Head|•>]  [7|•>]  [6|•>]  [4|•>]  [0|•>]  [8|•>] [(tail)Null]

	Showing your list ... There are  5  nodes in your singly-linked list.

	 ----------------------------------------



 	Let's PUSH a new Node with value of 999 for demo.
 [999|•>] 
	Confirm there are now there are  6 nodes.

	[Head|•>]  [999|•>]  [7|•>]  [6|•>]  [4|•>]  [0|•>]  [8|•>] [(tail)Null]





 	Let's POP a new Node ... removing the one closest to the head.

	Confirm there are now there are  6 nodes.

	[Head|•>]  [7|•>]  [6|•>]  [4|•>]  [0|•>]  [8|•>] [(tail)Null]




<pre>
import random
import numpy as np

#Singly-linked list 

class Node:
	def __init__(self, data, nextNode = None):
		self.data = data
		self.nextNode = nextNode
		
	def getData(self):
		return self.data
		
	def setData(self, val):
		self.data = val
		
	def getNextNode(self):
		return self.nextNode
		
	def setNextNode(self, val):
		self.nextNode = val
		
class SinglyLinkedList:
	listOfNodes = []  """ HERE IS THE LIST OF NODES """
	
	def __init__(self, head = None):
		self.head = head
		self.size = 0
		
	def getSize(self):
		return self.size
		
	def addNode(self, data, isPush = False):
		newNode = Node(data, self.head)
		self.head = newNode
		print(" ["+ str(data) + "|•>] ",end="")
		self.size += 1
		if isPush:
			self.listOfNodes.insert(0, newNode)
		else:
			self.listOfNodes.append(newNode)
		return True
		
	def printNode(self):
		curr = self.head
		while curr:
			curr = curr.getNextNode()

	def popNode(self):
		self.listOfNodes.pop(0)

""" --------------------------------------------------------- """
def main():
	
	 create a random list of values to create new nodes
	l = random.sample(range(0, 10), 5)

	# -------- create the sll ---------
	listDemo1 = SinglyLinkedList()

	# create the nodes ...
	print("[Head|•>]", end="")
    
	for i in l:
		listDemo1.addNode(i)
        
	print("[(tail)Null")

	print("Showing your list ... There are ",listDemo1.getSize()," nodes in your singly-linked list.\n")
	print("Let's PUSH a new Node with value of 999 for demo.")
	listDemo1.addNode(999, True)

	print("Confirm there are now there are ", listDemo1.getSize(), "nodes.")

	# convert our list of nodes back into Nodes themselves to use their methods.
	length = len(listDemo1.listOfNodes)
	print("[Head|•>] ", end="")
	for i in range(length):
		tempNode = listDemo1.listOfNodes[i]
		#print(tempNode.getData())
		print(" ["+ str(tempNode.getData())+ "|•>] ",end="")
	print("[(tail)Null"]\n\n")
		

	### -----------------------------------------------------
	# POP A NODE FROM THE HEAD 
	print("\n"*3, "\tLet's POP a Node ... removing the one closest to the head.")
	listDemo1.popNode()

	print("\n\tConfirm there are now there are ", listDemo1.getSize(), "nodes.")

	# convert our list of nodes back into Nodes themselves to use their methods.
    # for each node entry in the list, I can grab it, convert it back to Node, and get the Data...
    
	length = len(listDemo1.listOfNodes)
	print("\n\t[Head|•>] ", end="")
    
	for i in range(length):
		tempNode = listDemo1.listOfNodes[i]
		#print(tempNode.getData())
		print(" ["+ str(tempNode.getData())+ "|•>] ",end="")
          
	print("[(tail)Null"]\n\n")
	


if __name__ == "__main__":
	main()
</pre>

<hr />
<h2>Counting Frequencies</h2>
<p>Counting Frequencies is important and there are lots of ways to do so ... </p>

In [ ]:
# counting frequencies is done a lot of ways!
def count_the_frequencies(inputstr):
  
    # break the string into list of words
    list_of_terms = inputstr.split()
  
    # gives set of unique words
    unique_words = set(list_of_terms)
      
    for words in unique_words:
        print(words,":", list_of_terms.count(words))
  
# start here - parse a George Washington speech
if __name__ == "__main__":
      
    str = "While I give you these assurances, and pledge myself in the most unequivocal manner, to exert whatever ability I am possessed of, in your favor—let me entreat you, Gentlemen, on your part, not to take any measures, which, viewed in the calm light of reason, will lessen the dignity, & sully the glory you have hitherto maintained—let me request you to rely on the plighted faith of your Country, and place a full confidence in the purity of the intentions of Congress…"
      
    # calling the freq function
    count_the_frequencies((str.lower()))

In [ ]:
def count(elements):
    # check if each word has '.' at its last. If so then ignore '.'
    if elements[-1] == '.':
        elements = elements[0:len(elements) - 1]
   
    # if there exists a key as "elements" then simply
    # increase its value.
    if elements in dictionary:
        dictionary[elements] += 1
   
    # if the dictionary does not have the key as "elements" 
    # then create a key "elements" and assign its value to 1.
    else:
        dictionary.update({elements: 1})
   
   
# driver input to check the program.
   
str = "While I give you these assurances, and pledge myself in the most unequivocal manner, to exert whatever ability I am possessed of, in your favor—let me entreat you, Gentlemen, on your part, not to take any measures, which, viewed in the calm light of reason, will lessen the dignity, & sully the glory you have hitherto maintained—let me request you to rely on the plighted faith of your Country, and place a full confidence in the purity of the intentions of Congress…"
   
# Declare a dictionary
dictionary = {}
   
# split all the word of the string.
lst = str.split()
   
# take each word from lst and pass it to the method count.
for elements in lst:
    count(elements)
   
# print the keys and its corresponding values.
print("Frequencies ... ")
for allKeys in dictionary:
    print(allKeys, end = " ")
    print(":", end = " ")
    print(dictionary[allKeys], end = " ")
    print() 

In [ ]:
from collections import Counter

inputstr = "While I give you these assurances, and pledge myself in the most unequivocal manner, to exert whatever ability I am possessed of, in your favor—let me entreat you, Gentlemen, on your part, not to take any measures, which, viewed in the calm light of reason, will lessen the dignity, & sully the glory you have hitherto maintained—let me request you to rely on the plighted faith of your Country, and place a full confidence in the purity of the intentions of Congress…"
list_of_terms = inputstr.split()

print("-"*40,"\nVersion 1\n")
cnt = Counter()
for word in list_of_terms: 
    cnt[word] += 1
print(cnt)

# second version
print("\n\n","-"*40,"\nVersion 2 - with dictionary \n")
dictionaryVersion = dict()

for word in list_of_terms:
    if word in dictionaryVersion:
        dictionaryVersion[word] = dictionaryVersion[word] + 1
    else: # add to dictionary and count 1
        dictionaryVersion[word] = 1
    
# show our new frequencies from dictionary
for key in list(dictionaryVersion.keys()):
    print(key, ":", dictionaryVersion[key])
    
print("\n\nSorted by count\n")
{key: value for key, value in sorted(dictionaryVersion.items(), key = lambda item: item[1])}